### ISSUE TITLE
[New build - DEFAULT] 20w30c: Player Locomotion Improvements

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#2704]
Various fixes and improvements relative to accessibility, player locomotion and occlusion culling.

Changes
- Loading time should be a bit faster.
- Unstuck Mode added: press F5 to activated it.
- Is stuck message.
- Command List Menu: list all "hidden" hard keybinds like the F keys.
- Player stops walking faster.
- Fast Speed key are now a debug toggle.
- Fast Speed factor HUD indicator.

Bugfixes:
#2690 Wrong maglock mass limit computation.
Player will be less likely to get stuck after a high fall.
Inconsistent player jump.
#2694 Maglock: Parented entity issue when parent is unspawned.
#2700 Occlusion culling can cull close bricks. 
Occlusion culling can cull brick behind two closer bricks.
Some decals are not visible up-close.

**Comment by tsunamayo** [comment]
Build is live!

**Comment by Kaiser-Indrasil** [comment]
I love the inclusion of the Unstuck Mode - it is a godsend when building in a cramped, hard-to-get-to area. Thank you!

In other news, there is a bug where my character can't get through 2m-tall doors, like the ones in the Tutorial Building 1.
Also, looks like #2238 is back.

**Comment by ProPeach** [comment]
The new no-clip/unstuck mode is awesome! Walking now feels much tighter too. I was actually going to suggest that the new Function key uses are shown in the Keybind settings, but you've already thought of that with the Command List!

Can I request that the Fast Mode be removed from the Debug menu, and constantly available to us in Creative mode like it was previously? I use it literally every minute while playing, it's just so amazingly handy when building or moving between ships in the huge gameworld. I worry that new players will get frustrated with the slow default speed and not know about the optional Fast Mode because they haven't checked the Commands List. The new indicator that shows the speed multiplier is awesome when it is activated with F3, but if it showed all the time then it would inform newer players that the Fast Mode is available to them, and encourage them to look for the keybind to increase their speed.

I know there are some bugs with the Fast Mode if you go too fast at the edge of the galaxy, but I think players will understand that going 14 million times the default speed is the cause of that issue.

**Comment by BrickMacklin** [comment]
Agreed with Peach. The movement toggle is an absolute necessity while building. Keeping it as a default for creative is a satisfactory compromise.

**Comment by Vicomt** [comment]
Agreed with Peach and Brick - large builds absolutely require moving around quickly, and hiding it in a debug option is IMHO the wrong way to do it - I'd prefer to see it as a keybind/control that only works when in creative mode.

**Comment by ZachZent** [comment]
Here is a possible compromise for future survival. Have the fast mode unlocked everywhere in creative mode and only in shipyards in survival. Same with any other movement option that makes the building easier but would be a bit cheaty in survival like no clipping. That way players will have the same ability to build in survival and creative and thus not feel it is hampered when building in survival, hopping to creative mode to create an awesome ship to use in survival. Players can build anywhere in survival, but only have the full movement options inside shipyards. Either the player simply goes inside of the boundaries for access to these option in survival or a sort of interface linked with the shipyard that puts you in a build mode.

**Comment by Rexxios** [comment]
The AI is still non aggressive for me, has there been any news on this?

**Comment by tsunamayo** [comment]
yes #2238 it is actually super tricky, this for example was messing up with jump speed so I will have to find another way of doing it.
For the fast mode I want to find a smarter way of solving that issue. How is it working in other game (my classical question lol)

**Comment by Kaiser-Indrasil** [comment]
Well, no other game I've played actually uses something like a fast movement mode but with how useful it is, I'm surprised they don't.

I think Zach's solution is an elegant one but I would top it off with some troll-prevention mechanics, so that other players can't modify our build unless we whitelist them. Rust does it by letting players spawn a wardrobe (sic!) inside their bases which acts exactly like a whitelist, barring all other players from putting down anything inside the wardrobe's sphere of influence.
In the scenario where players build bases in multiplayer server, this could prevent griefing in Starship Evo too.

A similar solution would be to just introduce a similar block that would whitelist players to modify a given grid instead of sphere.

**Comment by ZachZent** [comment]
> I think Zach's solution is an elegant one but I would top it off with some troll-prevention mechanics, so that other players can't modify our build unless we whitelist them.

This is what I was thinking of when I said "a sort of interface linked with the shipyard that puts you in a build mode." You link this terminal to a shipyard and only those in the computer can access the shipyard. Maybe the player is represented with a drone that can't leave the shipyard boundaries (*cough* FR-3N/player made drones *cough*.) Additional permissions can be added to the terminal in time depending on who is using it. Solo only, group, open, etc.

Also it would allow players to link it in inside a station and build in a shipyard some distance away then take a shuttle to get in the ship. Like one of those ship reveal scenes every single star trek show has at some point. Most of the movies too.

**Comment by tsunamayo** [comment]
I was thinking for doing something with the SHIFT key, to boost it a bit when you are airborne in anti-grav or 6dof.

**Comment by tsunamayo** [comment]
@Rexxios are you talking about #2657? Thanks

**Comment by BrickMacklin** [comment]
I'd hate to not use the brackets. When I was new that was not daunting at all.

**Comment by ProPeach** [comment]
Like an increase to the speed we currently get by pressing shift?
I'm not sure that it could be adapted to replace the brackets as you really need the fine control of choosing the speed multiplier yourself. What speed you choose is very dependant on what size ship you're working on, or how far you want to travel.

**Comment by Rexxios** [comment]
@tsunamayo Yeah, I would love to have a fight in single player :)

**Comment by tsunamayo** [comment]
@Rexxios it works for me, could you send me your blueprint? Thanks

**Comment by Rexxios** [comment]
[TIE-Ln Fighter Blueprint.zip](https://github.com/tsunamayo/Starship-EVO/files/4978186/TIE-Ln.Fighter.Blueprint.zip)
@tsunamayo
